File: host/src/main/java/dk/dtu/domain/program/ProgramCard.java
package dk.dtu.domain.program;

import dk.dtu.domain.core.reaction.ReactionKind;
import java.util.List;

/**
 * @author Weihao Mo
 * @author William Pii Jæger
 */
public record ProgramCard(Action action, int steps) {
    /**
     * @author William Pii Jæger
     * @author Weihao Mo
     */
    public ProgramCard {
        switch (action) {
            case MOVE -> {
                if (steps < -1 || steps > 3) {
                    throw new IllegalArgumentException("MOVE step must be -1..3");
                }
            }
            case ROTATERIGHT, ROTATELEFT, UTURN,SPAM,TROJAN_HORSE,WORM,AGAIN -> steps = 0;
            case SANDBOX, WEASEL, SPEED -> steps = 0;
        }
    }

    /**
     * @author William Pii Jæger
     * @author Weihao Mo
     */
    public List<ProgramOP> toOps() {
        return switch (action) {
            case MOVE      -> List.of(new ProgramOP.Move(steps));
            case ROTATERIGHT -> List.of(new ProgramOP.RotateRight());
            case ROTATELEFT  -> List.of(new ProgramOP.RotateLeft());
            case UTURN       -> List.of(new ProgramOP.UTurn());
            case SPAM -> List.of(new ProgramOP.Spam());
            case TROJAN_HORSE -> List.of(new ProgramOP.TrojanHorse());
            case WORM -> List.of(new ProgramOP.Worm());
            case SANDBOX -> List.of(new ProgramOP.Reaction(ReactionKind.SANDBOX));
            case WEASEL -> List.of(new ProgramOP.Reaction(ReactionKind.WEASEL));
            case SPEED -> List.of(new ProgramOP.Reaction(ReactionKind.SPEED));
            case AGAIN -> List.of(new ProgramOP.Again());
        };
    }

    /**
     * @author Weihao Mo
     * @author Bjarke Søderhamn Petersen
     * @author Asger Allin Jensen
     */
    public ProgramOP toOp() {
        return switch (action) {
            case MOVE      -> new ProgramOP.Move(steps);
            case ROTATERIGHT -> new ProgramOP.RotateRight();
            case ROTATELEFT  -> new ProgramOP.RotateLeft();
            case UTURN       -> new ProgramOP.UTurn();
            case SPAM -> new ProgramOP.Spam();
            case TROJAN_HORSE -> new ProgramOP.TrojanHorse();
            case WORM -> new ProgramOP.Worm();
            case SANDBOX -> new ProgramOP.Reaction(ReactionKind.SANDBOX);
            case WEASEL -> new ProgramOP.Reaction(ReactionKind.WEASEL);
            case SPEED -> new ProgramOP.Reaction(ReactionKind.SPEED);
            case AGAIN -> new ProgramOP.Again();

        };
    }

    /**
     * @author William Pii Jæger
     * @author Weihao Mo
     */
    @Override
    public String toString() {
        return switch (action) {
            case MOVE -> "MOVE" + steps;
            case ROTATELEFT -> "ROTATELEFT";
            case ROTATERIGHT -> "ROTATERIGHT";
            case UTURN -> "UTURN";
            case SPAM -> "SPAM";
            case TROJAN_HORSE -> "TROJAN_HORSE";
            case WORM -> "WORM";
            case SANDBOX -> "SANDBOX";
            case WEASEL -> "WEASEL";
            case SPEED -> "SPEED";
            case AGAIN -> "AGAIN";
        };
    }

    public static ProgramCard move1() { return new ProgramCard(Action.MOVE, 1); }
    public static ProgramCard move2() { return new ProgramCard(Action.MOVE, 2); }
    public static ProgramCard move3() { return new ProgramCard(Action.MOVE, 3); }
    public static ProgramCard back1() { return new ProgramCard(Action.MOVE, -1); }
    public static ProgramCard right() { return new ProgramCard(Action.ROTATERIGHT, 0); }
    public static ProgramCard left()  { return new ProgramCard(Action.ROTATELEFT, 0); }
    public static ProgramCard uturn() { return new ProgramCard(Action.UTURN, 0); }
    public static ProgramCard spam() {return new ProgramCard(Action.SPAM, 0);}
    public static ProgramCard trojanHorse() {return new ProgramCard(Action.TROJAN_HORSE, 0);}
    public static ProgramCard worm() {return new ProgramCard(Action.WORM, 0);}

    public enum Action { MOVE, ROTATERIGHT, ROTATELEFT, UTURN, SPAM, TROJAN_HORSE,WORM,SANDBOX, WEASEL, SPEED, AGAIN}

    public static ProgramCard sandbox() { return new ProgramCard(Action.SANDBOX, 0); }
    public static ProgramCard weasel() { return new ProgramCard(Action.WEASEL, 0); }
    public static ProgramCard speed() { return new ProgramCard(Action.SPEED, 0); }
    public static ProgramCard again() {
        return new ProgramCard(Action.AGAIN, 0);
    }
}

File: engine/src/program/programCard.ts
import {
  ProgramOP,
  MoveOp,
  RotateRightOp,
  RotateLeftOp,
  UTurnOp,
  SpamOp,
  TrojanHorseOp,
  WormOp,
  AgainOp,
  ReactionOp,
} from "./programOp.js";

/** Ported from dk.dtu.domain.program.ProgramCard.Action. */
export const Action = {
  MOVE: "MOVE",
  ROTATERIGHT: "ROTATERIGHT",
  ROTATELEFT: "ROTATELEFT",
  UTURN: "UTURN",
  SPAM: "SPAM",
  TROJAN_HORSE: "TROJAN_HORSE",
  WORM: "WORM",
  SANDBOX: "SANDBOX",
  WEASEL: "WEASEL",
  SPEED: "SPEED",
  AGAIN: "AGAIN",
} as const;

export type Action = (typeof Action)[keyof typeof Action];

/**
 * Ported from dk.dtu.domain.program.ProgramCard.
 * MOVE steps must be in -1..3; every non-MOVE action forces steps to 0.
 */
export class ProgramCard {
  readonly action: Action;
  readonly steps: number;

  constructor(action: Action, steps: number) {
    this.action = action;
    if (action === Action.MOVE) {
      if (steps < -1 || steps > 3) {
        throw new Error("MOVE step must be -1..3");
      }
      this.steps = steps;
    } else {
      this.steps = 0;
    }
  }

  toOps(): ProgramOP[] {
    return [this.toOp()];
  }

  toOp(): ProgramOP {
    switch (this.action) {
      case Action.MOVE:
        return new MoveOp(this.steps);
      case Action.ROTATERIGHT:
        return new RotateRightOp();
      case Action.ROTATELEFT:
        return new RotateLeftOp();
      case Action.UTURN:
        return new UTurnOp();
      case Action.SPAM:
        return new SpamOp();
      case Action.TROJAN_HORSE:
        return new TrojanHorseOp();
      case Action.WORM:
        return new WormOp();
      case Action.SANDBOX:
        return new ReactionOp("SANDBOX");
      case Action.WEASEL:
        return new ReactionOp("WEASEL");
      case Action.SPEED:
        return new ReactionOp("SPEED");
      case Action.AGAIN:
        return new AgainOp();
    }
  }

  toString(): string {
    if (this.action === Action.MOVE) return "MOVE" + this.steps;
    return this.action;
  }

  static move1(): ProgramCard {
    return new ProgramCard(Action.MOVE, 1);
  }
  static move2(): ProgramCard {
    return new ProgramCard(Action.MOVE, 2);
  }
  static move3(): ProgramCard {
    return new ProgramCard(Action.MOVE, 3);
  }
  static back1(): ProgramCard {
    return new ProgramCard(Action.MOVE, -1);
  }
  static right(): ProgramCard {
    return new ProgramCard(Action.ROTATERIGHT, 0);
  }
  static left(): ProgramCard {
    return new ProgramCard(Action.ROTATELEFT, 0);
  }
  static uturn(): ProgramCard {
    return new ProgramCard(Action.UTURN, 0);
  }
  static spam(): ProgramCard {
    return new ProgramCard(Action.SPAM, 0);
  }
  static trojanHorse(): ProgramCard {
    return new ProgramCard(Action.TROJAN_HORSE, 0);
  }
  static worm(): ProgramCard {
    return new ProgramCard(Action.WORM, 0);
  }
  static sandbox(): ProgramCard {
    return new ProgramCard(Action.SANDBOX, 0);
  }
  static weasel(): ProgramCard {
    return new ProgramCard(Action.WEASEL, 0);
  }
  static speed(): ProgramCard {
    return new ProgramCard(Action.SPEED, 0);
  }
  static again(): ProgramCard {
    return new ProgramCard(Action.AGAIN, 0);
  }
}
